### DISCUSSION TITLE
Why does general `jnp.arange` calls materialise the array by default?

### DISCUSSION THREAD

**Discussion by soraros** [jax-ml/jax#23429] [General]
When `step is None`, a `lax.iota` is generated. However, `jnp.arange(0, 10, 2)` will become a constant. Why isn't it implemented as `iota() * step + start`?

**Comment by jakevdp** [comment]
I think it’s mainly because floating point rounding errors often over-shoot or under-shoot the endpoint, so that the strategy you suggest causes the length of the output array to be off by one compared to what the user expects. That said, in the integer case we could probably avoid materialization
